### PR TITLE
add subgraph support

### PR DESCRIPTION
**Pull request by yuliying**
add subgraph support

### CODE REVIEW

File: graphviz.lua [hunk 1]
```diff
@@ -35,20 +40,50 @@ local __Graph = {
...
+		end
+
+		local lines = split(src , "\n")
+		local indentstr = string.rep("\t\t" , level)
+		src = indentstr .. table.concat(lines , "\n" .. indentstr)
+
+		return src
 	end,
 
+	addsubgraph = function(self , subgraphname)
```
Comment (Nymphium) [inline graphviz.lua]: method name should be `subgraph`, like `edge`, `node`.

File: graphviz.lua [hunk 2]
```diff
@@ -35,20 +40,50 @@ local __Graph = {
...
+		local subgraph = Graph()
+		self.__subgraphs[subgraphname] = subgraph
+		return subgraph
+	end ,
```
Comment (Nymphium) [inline graphviz.lua]: remove space  between`end`  `,`

File: graphviz.lua [hunk 3]
```diff
@@ -124,16 +159,17 @@ local style_index = {
...
-		 edges = {
+		edges = {
 			 edge = {},
 			 style = setmetatable({}, {__index = style_index})},
 		nodes = {
 			node = {},
 			style = setmetatable({}, {__index = style_index})},
 		graph = {
 			style = setmetatable({}, {__index = style_index})},
+		__subgraphs = {} ,
```
Comment (Nymphium) [inline graphviz.lua]: field name should be `subgraphs`, like `edges` and `nodes`.

File: graphviz.lua [hunk 4]
```diff
@@ -35,20 +40,50 @@ local __Graph = {
...
+
 		local edge = self.edges.edge
 		for i = 1, #edge do
 			src = src .. ("\t\t\t%s -> %s\n"):format(edge[i].prev, edge[i].succ)
 		end
 
-		return src .. "}"
+		src = src .. "}"
+
+		--indent
```
Comment (Nymphium) [inline graphviz.lua]: in this case `delimiter` is fixed by `"\n"` so the argument is not needed.
More to say, this `split` should be like this:

```lua
local function split(str)
	if not str:match("\n$") then
		str = str .. "\n"
	end

	local ret = {}

	for w in str:gmatch("(.-)\n") do
		table.insert(ret, w)
	end

	return ret
end

local lines = split(src)
......
```

**Review by Nymphium — CHANGES_REQUESTED** [review]

File: graphviz.lua (hunk 4 above)
Comment (yuliying) [reply graphviz.lua]: yes , this split functions looks great.

**Comment by yuliying** [comment]
all you mentioned fixed .

File: graphviz.lua [hunk 5]
```diff
@@ -53,32 +53,32 @@ local __Graph = {
 		src = src .. "}"
 
 		--indent
```
Comment (Nymphium) [inline graphviz.lua]: remove unneeded comment

File: graphviz.lua [hunk 6]
```diff
@@ -169,7 +169,7 @@ Graph = function()
 			style = setmetatable({}, {__index = style_index})},
 		graph = {
 			style = setmetatable({}, {__index = style_index})},
-		__subgraphs = {} ,
+		subgraphs = {} ,
```
Comment (Nymphium) [inline graphviz.lua]: remove a space between `{}` and `,`

**Review by Nymphium — CHANGES_REQUESTED** [review]
finally please revise this

**Comment by Nymphium** [comment]
:+1:

**Pull request merged**